ouroboros-network/src/Ouroboros/Network/Subscription/PeerState.hs

```diff
@@ -253,6 +256,69 @@ instance Eq addr
 newPeerStatesVar :: MonadSTM m => m (StrictTVar m (PeerStates m addr))
 newPeerStatesVar = newTVarM (PeerStates Map.empty)
 
+
+-- | Periodically clean 'PeerState'.  It will stop when 'PeerState' becomes
+-- 'ThrowException'.
+--
+cleanPeerStates :: ( MonadSTM   m
+                   , MonadAsync m
+                   , MonadTime  m
+                   , MonadTimer m
+                   )
+                => DiffTime
+                -> StrictTVar m (PeerStates m addr)
+                -> m ()
+cleanPeerStates interval v = go
+  where
+    go = do
+      threadDelay interval
```

Comment:
Do we have to do this with polling? Can't we do use STM to wait for the state to change?

Reply:
We could, but we'll need to use `unsafeIOToSTM`.  We need `IO` to read the current time with `getMonotonicTime`.  We use an unsafe foreign call `getMonotonicNSec` so I think it's safe to wrap it in `unsafeIOToSTM`:
```
forever $ atomically $ do
  s <- readTVar v
  t <- unsafeIOToSTM getMonotonicTime
  ...
```